awpy/visualization/plot.py

```diff
@@ -326,3 +341,382 @@ def plot_nades(
+        given_fig_ax (tuple): Fig and ax objects if given
+
+    Returns: Plot given map control metric values onto axes object
+
+    Raises:
+        ValueError: If metrics is empty
+    """
+    if not metric_arr:
+        msg = "Metrics is empty."
+        raise ValueError(msg)
+
+    if given_fig_ax[0] is not None:
+        _plot_map_control_metrics_helper(metric_arr, given_fig_ax[1])
+    else:
+        _, curr_ax = plt.subplots()
+        _plot_map_control_metrics_helper(metric_arr, curr_ax)
+        plt.show()
+
+
+def _save_map_control_graphic_helper(
+    map_name: str,
+    frame: GameFrame,
+    metrics: list[float],
+    save_path: str = "",
+) -> None:
+    """Helper function to generate map control graphic for awpy round.
+
+    Args:
+        map_name (str): Map used in frame_map_control_metric call
+        frame (GameFrame): awpy frame to create map control graphic
+        metrics  (list): List containing map control values to plot
+        save_path (str): Location where graphic should be saved on file
+
+    Raises:
+        ValueError: If map_name is not in awpy.data.NAV
+    """
+    if map_name not in NAV:
+        msg = "Map not found."
+        raise ValueError(msg)
+
+    fig, axs = plt.subplots(1, 2, figsize=(12, 12))
+    map_bg = plt.imread(f"../awpy/data/map/{map_name}_dark.png")
+    axs[0].imshow(map_bg, zorder=0)
+    plot_frame_map_control(
+        map_name, frame, plot_type="players", given_fig_ax=(fig, axs[0])
+    )
+    _plot_map_control_metrics_helper(metrics, axs[1])
+    plt.tight_layout()
+    if save_path:
+        fig.savefig(fname=save_path, bbox_inches="tight", dpi=400)
+    else:
+        plt.show(fig)
+    plt.close(fig)
+
+
+def save_map_control_graphic(
+    map_name: str,
+    frames: list[GameFrame],
+    save_path: str = "",
+) -> None:
+    """Function to generate map control gif for awpy round.
+
+    Gif including minimap visualization
+    and map control metric plot for each frame in the round.
+
+    Args:
+        map_name (str): Map used inframe_map_control_metric call
+        frames (list): List of GameFrame for map control graphic
+        save_path (str): Location where graphic should be saved on file
+
+    Raises:
+        ValueError: If map_name is not in awpy.data.NAV
+    """
+    if map_name not in NAV:
+        msg = "Map not found."
+        raise ValueError(msg)
+
+    metrics: list[float] = []
+    images: list[np.ndarray] = []
+
+    print("Saving/loading frames!")
+    for i, frame in enumerate(frames):
+        metrics.append(frame_map_control_metric(map_name, frame))
+        temp_filename = f"./tmp/tmp{i!s}.png"
+        _save_map_control_graphic_helper(
+            map_name, frame, metrics, save_path=temp_filename
+        )
+        if save_path:
+            # sometime second dimension has an extra entry
+            # so resize to expected size
+            cur_img = cv2.resize(imageio.imread(temp_filename), (4760, 4760))
+            images.append(cur_img)
```

Comment:
Is the problem here only ever that some images have exactly one extra entry somewhere?

Couldnt you just work on the numpy arrays directly instead of needing that heavy dependency?

Also do you always want exactly that size or do you also want it configurable.

Reply:
Yeah there was a weird cropping effect on the temporary images that were saved when the gif was created. That cropping effect then made a weird jittering effect on the subsequent gif. I tried to get around the issue with numpy but I wasn't able to seem to get it working whenever the cropping effect would occur (which was not consistent). When I have all the changes working and addressing all the comments, I'll see if I can get numpy to work such that cv2 can be replaced as a dependency. I do think it's nice that cv2 is in the package now since its resize is the best way to resize images in case its needed in the future for the package but if I can remove it I'll try to.

Yeah I think it makes sense to make the size configurable. I'll make the default the size that is currently being used but I'll allow for other sizes as well.

I'll resolve this conversation once I actually make the change btw.

Reply:
Addressed specifying figure size if wanted in next commit on branch